src/core/graph.js

```diff
@@ -889,6 +898,63 @@ export class Graph {
     return result;
   }
 
+  /**
+   * Create a new graph, in which some nodes have been contracted together.
+   *
+   * contract takes a list of GraphContractions, each of which specifies a
+   * replacement node, and a list of old node addresses to map onto the new
+   * node. A new graph will be returned where the new node is added, none of
+   * the old nodes are present, and every edge incident to one of the old nodes
+   * has been re-written so that it is incident to the new node instead.
+   *
+   * If the same node addresses is "old" for several contractions, all incident
+   * edges will be re-written to connect to whichever contraction came last.
+   *
+   * If the replacement node is present in the graph, no error will be thrown,
+   * provided that the replacement node is consistent with the one in the graph.
+   *
+   * If there is a "chain" of remaps (i.e. a->b and b->c), then an error will
+   * be thrown, as support for chaining has not yet been implemented.
+   *
+   * The original Graph is not mutated.
+   *
+   * contract runs in O(n+e+k), where `n` is the number of nodes, `e` is the
+   * number of edges, and `k` is the number of contractions. If needed, we can
+   * improve the peformance by mutating the original graph instead of creating
+   * a new one.
+   */
+  contract(contractions: $ReadOnlyArray<GraphContraction>): Graph {
```

Comment:
nit: You might consider `contractNodes` and `NodeContraction`, as
without further qualification *contraction* generally refers to an
operation on a single edge, but it’s clear enough either way.

Reply:
Done.